joint_calling/upload_processor.py

```diff
@@ -0,0 +1,110 @@
+""" This function prepares gVCF's uploaded to GCS, based off sample status
+    logged in a database, for further QC and downstream analysis.
+    The upload processor will determine when samples should be added
+    to existing MatrixTables where appropriate and which MatrixTables
+    they should be combined with in this case. Following a successful
+    run all uploaded files will be moved to archival storage.
+
+    Assumptions
+    ===========
+    - Specified files must exist in either the source or the destination
+      bucket
+    - Source and destination bucket must exist
+    - User must be authenticated with appropriate permissions """
+
+import errno
+import os
+import subprocess
+from typing import List
+import hailtop.batch as hb
+
+
+def batch_move_files(
+    batch: hb.batch,
+    files: List[str],
+    source_prefix: str,
+    destination_prefix: str,
+    docker_image: str = None,
+    key: str = None,
+) -> List:
+    """This script takes a Hail.Batch workflow and a list of files to
+    move between buckets. It adds 1 job per file to the workflow, and
+    runs in a docker_image. It returns the list of created jobs. Each
+    job consists of a gsutil move command for each valid file.
+    When run, these jobs will perform the batch move file operation.
+
+    Parameters
+    ==========
+    batch: hb.Batch
+        An object representing the DAG of jobs to run.
+    files: List[str]
+        A list of the file names to be moved.
+        For example ["TOB1543.g.vcf.gz","TOB2314.g.vcf.gz","TOB3423.g.vcf.gz"]
+    source_prefix: str
+        The path to the sub-directory where the files are initially located.
+        For example "cpg-tob-wgs-upload" or "cpg-tob-wgs-upload/v1"
+    destination_prefix: str
+        The path to the sub-directory where the files should be moved.
+        For example "cpg-tob-wgs-main" or "cpg-tob-wgs-upload/batch0"
+    docker_image: str, optional
+        The address and tag of a previously built docker image, within the
+        artifact registry. Each batch job will run in this image.
+        For example;
+        australia-southeast1-docker.pkg.dev/project/images/driver:version'
+    key: str, optional
+        key-file for the service account used for authentication. In the case that this
+        is not provided as an input, it is assumed that this key will exist at
+        /gsa-key/key.json. This is the case when using the hail batch service backend.
+        For example:
+        "{
+            "type": "service_account",
+            "project_id": "",
+            "private_key_id": "" ...
+        }
+        " """
+
+    jobs = []
+
+    for sample in files:
+
+        previous_location = os.path.join('gs://', source_prefix, sample)
+        new_location = os.path.join('gs://', destination_prefix, sample)
+
+        # Checks if the files exist at the source and destination
+        get_file_source = subprocess.run(
```

Comment:
Instead of performing these checks before you _create_ the job, it would be more robust to perform them as part of _executing_ the job.

Consider that we want to move A --> B, with the following sequence of events:

1. A exists, B doesn't.
2. `batch_move_files` gets called, a job gets created.
3. Something _else_ (we don't expect that, but who knows what will happen!) moves A to B.
4. The batch gets run.
5. The batch fails, even though there actually is no problem.

This is very much an edge case, but if it's easy to move the checks to the job itself, it might be worth doing so. Your tests in that case would need to be able to check whether the batch executed successfully or not.

Reply:
Great point, have tried to implement this in the latest commit! Hopefully, it's suitable.